File: frontend/src/api/jobs.ts
import type { Job } from '../types'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://127.0.0.1:8000'

export type MarketInsightsResponse = {
  overview: {
    total_jobs: number
    remote_jobs: number
    hybrid_jobs: number
    onsite_jobs: number
  }
  trending_skills: { name: string; count: number }[]
  top_locations: { city: string; count: number }[]
  top_companies: { name: string; count: number }[]
}

export async function fetchJobs(): Promise<Job[]> {
  const response = await fetch(`${API_BASE_URL}/jobs/`)

  if (!response.ok) {
    if (response.status === 503) {
      throw new Error('The live job service is temporarily unavailable. Please try again shortly.')
    }
    throw new Error('Failed to fetch jobs.')
  }

  return response.json()
}

export async function fetchJob(jobId: number): Promise<Job> {
  const response = await fetch(`${API_BASE_URL}/jobs/${jobId}`)

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error('This job is no longer available.')
    }
    if (response.status === 503) {
      throw new Error('The live job service is temporarily unavailable. Please try again shortly.')
    }
    throw new Error('Failed to fetch job details.')
  }

  return response.json()
}

export async function fetchMarketInsights(): Promise<MarketInsightsResponse> {
  const response = await fetch(`${API_BASE_URL}/jobs/market-insights`)

  if (!response.ok) {
    if (response.status === 503) {
      throw new Error('Market insights are temporarily unavailable.')
    }
    throw new Error('Failed to fetch market insights.')
  }

  return response.json()
}

File: backend/api/jobs.py
from collections import Counter
import logging

from fastapi import APIRouter, HTTPException

try:
    from backend.database.queries import (
        fetch_all_jobs_from_db,
        fetch_job_by_id_from_db,
    )
except ImportError:
    try:
        from database.queries import (
            fetch_all_jobs_from_db,
            fetch_job_by_id_from_db,
        )
    except ImportError:
        fetch_all_jobs_from_db = None
        fetch_job_by_id_from_db = None

router = APIRouter(prefix="/jobs", tags=["jobs"])
logger = logging.getLogger(__name__)


def get_jobs_data():
    if fetch_all_jobs_from_db is None:
        logger.error("Jobs database query module is unavailable")
        raise HTTPException(status_code=503, detail="Job service temporarily unavailable")
    try:
        return fetch_all_jobs_from_db()
    except Exception:
        logger.exception("Failed to read jobs from the database")
        raise HTTPException(status_code=503, detail="Job service temporarily unavailable")


@router.get("/market-insights")
def get_market_insights():
    jobs = get_jobs_data()

    skill_counter = Counter()
    location_counter = Counter()
    company_counter = Counter()

    for job in jobs:
        for tag in job.get("tags", []):
            skill_counter[tag] += 1

        location = job.get("location", "Unknown")
        location_counter[location] += 1

        company = job.get("company", "Unknown")
        company_counter[company] += 1

    total_jobs = len(jobs)
    remote_jobs = sum(1 for job in jobs if job.get("hybrid") == "Remote")
    hybrid_jobs = sum(1 for job in jobs if job.get("hybrid") == "Hybrid")
    onsite_jobs = sum(1 for job in jobs if job.get("hybrid") == "On-site")

    return {
        "overview": {
            "total_jobs": total_jobs,
            "remote_jobs": remote_jobs,
            "hybrid_jobs": hybrid_jobs,
            "onsite_jobs": onsite_jobs,
        },
        "trending_skills": [
            {"name": name, "count": count}
            for name, count in skill_counter.most_common(6)
        ],
        "top_locations": [
            {"city": city, "count": count}
            for city, count in location_counter.most_common(5)
        ],
        "top_companies": [
            {"name": name, "count": count}
            for name, count in company_counter.most_common(5)
        ],
    }


@router.get("/")
def get_jobs():
    return get_jobs_data()


@router.get("/{job_id}")
def get_job_by_id(job_id: int):
    if fetch_job_by_id_from_db is None:
        logger.error("Jobs database query module is unavailable")
        raise HTTPException(status_code=503, detail="Job service temporarily unavailable")
    try:
        job = fetch_job_by_id_from_db(job_id)
    except Exception:
        logger.exception("Failed to read job id=%s from the database", job_id)
        raise HTTPException(status_code=503, detail="Job service temporarily unavailable")
    if job:
        return job

    raise HTTPException(status_code=404, detail="Job not found")
